logging/unit_tests/test__shutdown.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+import mock
+
+
+class Test_setup_shutdown_stracktrace_reporting(unittest.TestCase):
+
+    def _call_fut(self, request):
+        from google.cloud.logging._shutdown import (
+            setup_stacktrace_crash_report)
+        return setup_stacktrace_crash_report(request)
+
+    def test_setup_shutdown_stacktrace_reporting_no_gae(self):
+        with self.assertRaises(RuntimeError):
+            self._call_fut(mock.Mock())
+
+    def test_setup_shutdown_stacktrace_reporting(self):
+        from google.cloud.logging._environment_vars import (
+            _APPENGINE_FLEXIBLE_ENV_VM)
+        from google.cloud._testing import _Monkey
+        import os
+
+        signal_patch = mock.patch('google.cloud.logging._shutdown.signal')
+
+        with _Monkey(os, environ={_APPENGINE_FLEXIBLE_ENV_VM: 'True'}):
+            with signal_patch as signal_mock:
+                self._call_fut(mock.Mock())
+                self.assertTrue(signal_mock.signal.called)
+
+
+class Test_write_stackrace_log(unittest.TestCase):
+
+    def _call_fut(self, client, traces):
+        from google.cloud.logging._shutdown import (
+            _write_stacktrace_log
+        )
+        return _write_stacktrace_log(client, traces)
+
+    def test_write_stracktrace_log(self):
+        from google.cloud._testing import _Monkey
+        import os
+
+        mock_client = mock.Mock()
```

Comment:
Please **always** set a strict `spec` on your mocks. Start with `mock.Mock(spec=[])` and build up as needed. In this case, you also want to mock `client.logging_api` and `client.logging_api.write_entries` (being deliberate is important to **some degree**)